src/main/java/com/google/sps/servlets/DataFromDatabase.java

```diff
@@ -0,0 +1,134 @@
+        List<String> spannerTypes = new ArrayList<>();
+        List<String> nullables = new ArrayList<>();
+
+        while (resultSet.next()) {
+          colnames.add(resultSet.getString(0));
+          spannerTypes.add(resultSet.getString(1));
+          nullables.add(resultSet.getString(2));
+        }
+
+        String query = "SELECT ";
+        for (String colName : colnames) {
+          query += (colName + ", ");
+        }
+        query = query.substring(0, query.length() - 2);
+        query += " FROM " + table; 
+
+        response.getWriter().println("Table: " + table);
+
+        executeTableQuery(query, colnames, spannerTypes, response);
+      }
+    }
+  }
+
+  private void executeTableQuery(String query, List<String> colnames, List<String> spannerTypes, HttpServletResponse response) throws IOException {
+    try (ResultSet rs =
+      dbClient
+      .singleUse() // Execute a single read or query against Cloud Spanner.
+      .executeQuery(Statement.of(query))) {
+
+      while (rs.next()) {
+        String row = "";
+        int index = 0;
+        while (index < colnames.size()) {
+          String colName = colnames.get(index);
+          row += (" " + colName + ": ");
+
+          // If there is a null in this col here, just print out NULL for now.
+          if (rs.isNull(colName)) {
+            index++;
+            row += "NULL"; 
+            continue;
+          }
+
+          // Figure out how to make more concise.
+          switch (spannerTypes.get(index)) {
+            case "STRING(MAX)":
+              row += rs.getString(colName);
+              break;
+            case "STRING(250)":
+              row += rs.getString(colName);
+              break;
+            case "STRING(1024)":
+              row += rs.getString(colName);
+              break;
+            case "TIMESTAMP":
+              row += rs.getTimestamp(colName);
+              break;
+            case "INT64":
+              row += rs.getLong(colName);
+              break;
+            case "BYTES(MAX)":
+              ByteArray bytes = rs.getBytes(colName);
+              byte[] byteArray = bytes.toByteArray();
+          
+              for (byte b : byteArray) {
+                row += b;
+              }
+              break;
+            case "BYTES":
+              row += rs.getBytes(colName);
+              break;
+            case "ARRAY<INT64>":
+              row += "ARRAY HERE";
+              break;
+            case "DATE":
+              row += "DATE HERE";
+              break;
+            case "BOOL":
+              row += rs.getBoolean(colName);
+          }
+          index++;
+        }
+        response.getWriter().println(row); 
+      }
+    }
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    selectedTables = request.getParameterValues("table-select");
+    response.sendRedirect("/select-tables.html");
```

Comment:
Instead of doing a POST to set selectedTables and then a GET to read from them can we not pass the selected tables in the GET request?

NOTE: this will eventually run on a number of machines where requests could go to any machine. You are keeping server side state here and when it runs on multiple machines this will break since the second request could go to a whole different machine (or worse, someone else's post could have changed selectedTables).

Reply:
Done